org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorLanguage.java

```diff
@@ -0,0 +1,380 @@
+		}
+		XtextResource resource = (XtextResource) resourceSet.getResource(URI.createURI(getGrammarUri()), true);
+		if (resource.getContents().isEmpty()) {
+			throw new IllegalArgumentException("Couldn't load grammar for '" + getGrammarUri() + "'.");
+		}
+		if (!resource.getErrors().isEmpty()) {
+			LOG.error(resource.getErrors());
+			throw new IllegalStateException(
+					"Problem parsing '" + getGrammarUri() + "':\n" + Joiner.on("\n").join(resource.getErrors()));
+		}
+		Grammar grammar = (Grammar) resource.getContents().get(0);
+		validateGrammar(grammar);
+		initialize(grammar);
+		super.initialize(injector);
+	}
+
+	protected List<? extends IXtextGeneratorFragment> getImplicitFragments() {
+		List<ImplicitFragment> fragments = new ArrayList<>();
+		fragments.add(new ImplicitFragment());
+		return fragments;
+	}
+
+	public void initialize(Grammar grammar) {
+		this.grammar = grammar;
+		this.ruleNames = RuleNames.getRuleNames(grammar, true);
+	}
+
+	private void installIndex() {
+		ResourceDescriptionsData index = new ResourceDescriptionsData(Collections.emptyList());
+		List<Resource> resources = Lists.newArrayList(resourceSet.getResources());
+		for (Resource resource : resources) {
+			index(resource, resource.getURI(), index);
+		}
+		ResourceDescriptionsData.ResourceSetAdapter.installResourceDescriptionsData(resourceSet, index);
+	}
+
+	private void index(Resource resource, URI uri, ResourceDescriptionsData index) {
+		IResourceServiceProvider serviceProvider = IResourceServiceProvider.Registry.INSTANCE
+				.getResourceServiceProvider(uri);
+		if (serviceProvider != null) {
+			IResourceDescription resourceDescription = serviceProvider.getResourceDescriptionManager()
+					.getResourceDescription(resource);
+			if (resourceDescription != null) {
+				index.addDescription(uri, resourceDescription);
+			}
+		}
+	}
+
+	protected void validateGrammar(Grammar grammar) {
+		validateAllImports(grammar);
+		EValidator validator = EValidator.Registry.INSTANCE.getEValidator(XtextPackage.eINSTANCE);
+		if (validator != null) {
+			DiagnosticChain chain = new DiagnosticChain() {
+				@Override
+				public void add(final Diagnostic diagnostic) {
+					if (diagnostic.getSeverity() == Diagnostic.ERROR) {
+						String grammarName = "Validation Error in " + grammar.getName() + ": ";
+						if (diagnostic.getException() == null) {
+							throw new IllegalStateException(grammarName + diagnostic.getMessage());
+						} else {
+							throw new IllegalStateException(grammarName + diagnostic.getMessage(),
+									diagnostic.getException());
+						}
+					}
+				}
+
+				@Override
+				public void addAll(final Diagnostic diagnostic) {
+					add(diagnostic);
+				}
+
+				@Override
+				public void merge(final Diagnostic diagnostic) {
+					throw new UnsupportedOperationException();
+				}
+			};
+			validator.validate(grammar, chain, null);
+			final TreeIterator<EObject> iterator = grammar.eAllContents();
```

Comment:
Cosmetics: unnecessary final. Could be Iterator instead of TreeIterator. Could be for-loop over EcoreUtil2.eAll(grammar).